src/main/thrift/parquet.thrift

```diff
@@ -835,6 +864,65 @@ struct ColumnMetaData {
   16: optional SizeStatistics size_statistics;
 }
 
+struct ColumnChunkMetaDataV3 {
```

Comment:
Unless I'm misunderstanding something, there should be one `FileColumnMetadataV3` element per leaf `SchemaElement`.

Reply:
> readers have to do a O(n) search to find a given column?

Why would they need an O(n) search? The index indexes an `SchemaElement[]` in java or `std::vector<SchemaElement>` in C++ which is O(1).

Reply:
Ok, let's step back because I think I am not following you.

Here is the file organization as proposed here:
1) `FileMetaDataV3` points to one `FileColumnMetadataV3` per column
2) `FileColumnMetadataV3` points to one `ColumnChunkV3` per row group

I'm not sure what you're proposing exactly when you mean "can skip encoding columns that do not have values"? This is currently not possible using Parquet, as at least the corresponding definition levels would be expected (even if they are all 0).

Reply:
> Well, ignoring the fact that Parquet is currently not a sparse format, your proposal implies that readers have to do a O(n) search to find a given column?

IIUC, Finding a column via schema elements today is also O(N) assuming no nesting.  I think the difference is today the first thing implementations do create an efficient dictionary structure to amortize lookup of further columns.  

I think if we want fast lookups without building any additional dictionaries in memory we should be considering a new stored index structure (or reconsider how we organize schema elements instead of a straight BFS).